Modules/CalcPerform.lua

```diff
@@ -14,9 +14,308 @@ local m_ceil = math.ceil
+			s_format(""),
+			s_format("%.3f ^8(adjusted for server tick rate)", rateCapAdjusted),
+			s_format("^8(extra ICDR of %d%% would reach next breakpoint)", extraICDRNeeded),
+			s_format(""),
+			s_format("Trigger rate:"),
+			s_format("1 / %.3f", rateCapAdjusted),
+			s_format("= %.2f ^8per second", 1 / rateCapAdjusted),
+		}
+	end
+	return 1 / rateCapAdjusted
+end
+
+-- Calculate Trigger Rate impact due to other skills in rotation
+-- This is achieved by simulation a 10 second cast rotation
+local function calcMultiSpellRotationImpact(env, skillRotation, sourceAPS)
+	local SIM_TIME = 100.0
+	local index = 1
+	local time = 0
+	local next_trigger = 0
+	local trigger_increment = 1 / sourceAPS
+	local wasted = 0
+
+	while time < SIM_TIME do
+		local currIndex = index
+	
+		if time >= next_trigger then
+			while skillRotation[index].next_trig > time do
+				index = (index % #skillRotation) + 1
+				if index == currIndex then
+					wasted = wasted + 1
+					break
+				end
+			end
+
+			if skillRotation[index].next_trig <= time then
+				skillRotation[index].count = skillRotation[index].count + 1
+				skillRotation[index].next_trig = skillRotation[index].next_trig + skillRotation[index].cd
+				index = (index % #skillRotation) + 1
+				next_trigger = next_trigger + trigger_increment
+			end
+		end
+		-- increment time by server tick time passage
+		time = time + (1 / data.misc.ServerTickRate)
+	end
+
+	local mainRate = 0
+	local trigRateTable = { simTime = SIM_TIME, rates = {}, }
+	if wasted > 0 then
+		t_insert(trigRateTable.rates, { name = "Wasted_Need Higher_ICDR", rate = wasted / SIM_TIME })
+	else
+		t_insert(trigRateTable.rates, { name = "No Wasted Hits_ _ ", rate = wasted / SIM_TIME })
+	end
+	for _, sd in ipairs(skillRotation) do
+		if cacheSkillUUID(env.player.mainSkill) == sd.uuid then
+			mainRate = sd.count / SIM_TIME
+		end
+		t_insert(trigRateTable.rates, { name = sd.uuid, rate = sd.count / SIM_TIME })
+	end
+	return mainRate, trigRateTable
+end
+
+-- Calculate the actual Trigger rate of active skill causing the trigger
+local function calcActualTriggerRate(env, source, sourceAPS, spellCount, output, breakdown, dualWield)
+	-- Get action trigger rate
+	if sourceAPS == nil and source.skillTypes[SkillType.Channelled] then
+		output.ActionTriggerRate = 1 / source.skillData.triggerTime
+		if breakdown then
+			breakdown.ActionTriggerRate = {
+				s_format("%.2f ^8(base cooldown of trigger)", source.skillData.triggerTime),
+				s_format(""),
+				s_format("Trigger rate:"),
+				s_format("1 / %.2f", source.skillData.triggerTime),
+				s_format("= %.2f ^8per second", output.ActionTriggerRate),
+			}
+		end
+	else
+		output.ActionTriggerRate = getTriggerActionTriggerRate(env.player.mainSkill.skillData.cooldown, env, breakdown)
+	end
+	local trigRate
+	local skillRotationImpact = #spellCount
+	if sourceAPS ~= nil then
+		output.SourceTriggerRate = sourceAPS / skillRotationImpact
+		if dualWield then
+			if #spellCount > 1 then
```

Comment:
The following two sections duplicate a lot of code.  Perhaps we can wrap the middle logic in a small function?